### PR TITLE
fixed kick event leave message

### LINKED ISSUE #4031: PlayerKickEvent leaveMessage is ignored
### What behaviour is expected:
When kicking a player, the set leaveMessage is used

### What behaviour is observed:
The default leave message is sent.

### Steps/models to reproduce:
`PlayerConnection#disconnect` throws a `PlayerKickEvent` which has a `leaveMessage` field.
`disconnect` then calls into `a(IChatBaseComponent)`, where the component is the quit reason written into the log files and displayed to the player
The actual quit message is defined by a call to `this.minecraftServer.getPlayerList().disconnect(this.player);`, because the `leaveMessage` is never passed into `a`.
 The Quit Message is then defined by the PlayerQuitEvent which is called by `PlayerList#disconnect`

### Plugin list:
none, observed by digging through the code

### Paper build number:
" 0b33230f2db4013d929b4b2e93b74ea57592a6d6 "

### Anything else:
Theoretically this bug should be resolved upstream, but I have low hopes that anyone in the chain is caring about this minor discrepancy. Some parties might even claim that everyone is relying on the current behavior.
The fix would be just to pass the leaveMessage to `a` for kicks and null for other disconnects.
I guess this issue is mostly there for people noticing this bug to confirm it's not an issue in their code.

### PR DESCRIPTION
**Pull request by Machine-Maker**
closes #4031

### CODE REVIEW

File: Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+         this.player.p();
+-        String quitMessage = this.minecraftServer.getPlayerList().disconnect(this.player);
+         if ((quitMessage != null) && (quitMessage.length() > 0)) {
+             this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage));
+         }
+@@ -1690,6 +1694,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
+         }
+ 
+     }
++    // Paper end
```
Comment (Proximyst) [inline Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch]: Move this up to the method declaration

File: Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
++
++    private void onAllDisconnect(IChatBaseComponent ichatbasecomponent, String quitMessage) {
+         // CraftBukkit start - Rarely it would send a disconnect line twice
+         if (this.processedDisconnect) {
+             return;
+@@ -1673,7 +1678,6 @@ public class PlayerConnection implements PacketListenerPlayIn {
+         */
+ 
+         this.player.p();
+-        String quitMessage = this.minecraftServer.getPlayerList().disconnect(this.player);
```
Comment (Proximyst) [inline Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch]: Comment this

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
++        // String quitMessage = this.minecraftServer.getPlayerList().disconnect(this.player); // Paper
+         if ((quitMessage != null) && (quitMessage.length() > 0)) {
+             this.minecraftServer.getPlayerList().sendMessage(CraftChatMessage.fromString(quitMessage));
+         }
+@@ -1688,7 +1693,7 @@ public class PlayerConnection implements PacketListenerPlayIn {
+             PlayerConnection.LOGGER.info("Stopping singleplayer server as player logged out");
+             this.minecraftServer.safeShutdown(false);
+         }
+-
++        // Paper end
```
Comment (Proximyst) [inline Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch]: I meant move this under the declaration, i.e.:
```java
private void onAllDisconnect(IChatBaseComponent ichatbasecomponent, String quitMessage) {
    // Paper end
```

**Review by Proximyst — DISMISSED** [review]
Minor issue; can be fixed on merge if you don't do it first :p

File: Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch (hunk 3 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0599-Fixed-kick-event-leave-message-not-being-sent.patch]: oh, ok I get it. I fixed it now.

**Comment by Machine-Maker** [comment]
`Rebased after the NMS repackage`

**Comment by Machine-Maker** [comment]
Looks like this also fixes [SPIGOT-3034](https://hub.spigotmc.org/jira/browse/SPIGOT-3034).

**Comment by Machine-Maker** [comment]
Rebased for 1.17.1

**Review by electronicboy — APPROVED** [review]

**Review by Proximyst — APPROVED** [review]

**Pull request merged**